workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/SpatialOSBehaviourManager.cs

```diff
@@ -1,60 +1,214 @@
         }
 
         public void EnableSpatialOSBehaviours()
         {
-            // inject + enable all spatialOSBehavioursToEnable
+            foreach (var behaviour in behavioursToEnable)
+            {
+                var dict = behaviourLibrary.InjectAllReadersWriters(behaviour, entity);
+                behaviourToReadersWriters[behaviour] = dict;
+                foreach (var idToReaderWriterList in dict)
+                {
+                    var id = idToReaderWriterList.Key;
+                    var readerWriterList = idToReaderWriterList.Value;
+                    foreach (var readerWriter in readerWriterList)
+                    {
+                        GetOrCreateValue(compIdToReadersWriters, id).Add(readerWriter);
+                    }
+                }
+            }
+
+            foreach (var behaviour in behavioursToDisable)
+            {
+                behaviour.enabled = true;
+            }
+
+            behavioursToEnable.Clear();
         }
 
         public void DisableSpatialOSBehaviours()
         {
-            // disable + deinject all spatialOSBehavioursToDisable
+            foreach (var behaviour in behavioursToDisable)
+            {
+                behaviour.enabled = false;
+            }
+
+            foreach (var behaviour in behavioursToDisable)
+            {
+                behaviourLibrary.DeInjectAllReadersWriters(behaviour);
+                foreach (var idToReaderWriterList in behaviourToReadersWriters[behaviour])
+                {
+                    var id = idToReaderWriterList.Key;
+                    var readerWriterList = idToReaderWriterList.Value;
+                    foreach (var readerWriter in readerWriterList)
+                    {
+                        compIdToReadersWriters[id].Remove(readerWriter);
+                    }
+                }
+
+                behaviourToReadersWriters.Remove(behaviour);
+            }
+
+            behavioursToDisable.Clear();
         }
 
         public void AddComponent(uint componentId)
         {
-            // Mark readers as to be enabled
+            // Mark reader components ready in relevant SpatialOSBehaviours
+            var relevantReaderSpatialOSBehaviours = behavioursRequiringReaderTypes[componentId];
+            MarkComponentRequirementSatisfied(relevantReaderSpatialOSBehaviours);
         }
 
         public void RemoveComponent(uint componentId)
         {
-            // Mark readers as to be disabled
+            // Mark reader components not ready in relevant SpatialOSBehaviours
+            var relevantReaderSpatialOSBehaviours = behavioursRequiringReaderTypes[componentId];
+            MarkComponentRequirementUnsatisfied(relevantReaderSpatialOSBehaviours);
         }
 
         public void ChangeAuthority(uint componentId, Authority authority)
         {
-            // Mark writers as to be enabled/disabled
+            if (authority == Authority.Authoritative)
```

Comment:
What about the third case when you transition from authoritative to authoritylossimminent?

Reply:
Discussed offline, this is OK since it's for lifecycle, not user callbacks.